src/lib/utils/utils.js

```diff
@@ -89,45 +89,93 @@ function getJson(url, cb) {
   httpGetJson(url, cb);
 }
 
-function pingEndpoint(host, port, type, protocol, origin, callback) {
-  const options = {
-    protocolVersion: 13,
-    perMessageDeflate: true,
-    origin: origin,
-    host: host,
-    port: port
-  };
-  if (type === 'ws') {
-    options.headers = {
-      'Sec-WebSocket-Version': 13,
-      Connection: 'Upgrade',
-      Upgrade: 'websocket',
-      'Sec-WebSocket-Extensions': 'permessage-deflate; client_max_window_bits',
-      Origin: origin
-    };
-  }
-  let req;
+function pingEndpoint(host, port, type, protocol, origin, callback, count = 0) {
   // remove trailing api key from infura, ie rinkeby.infura.io/nmY8WtT4QfEwz2S7wTbl
-  if (options.host.indexOf('/') > -1) {
-    options.host = options.host.split('/')[0];
-  }
-  if (protocol === 'https') {
-    req = require('https').get(options);
-  } else {
-    req = require('http').get(options);
-  }
+  const _host = host.indexOf('/') > -1 ? host.split('/')[0] : host;
+  const max_retries = 600;
+  const timeout = 1000;
 
-  req.on('error', (err) => {
-    callback(err);
-  });
+  let alreadyClosed = false;
+  let retrying = false;
 
-  req.on('response', (_response) => {
-    callback();
-  });
+  const doCallback = (v) => {
+    if (typeof v !== 'undefined') {
+      callback._doCallback = !!v;
+    }
+    let _doCallback = callback._doCallback;
+    if (typeof _doCallback === 'undefined') {
+      _doCallback = true;
+    } else {
+      _doCallback = !!_doCallback;
+    }
+    return _doCallback;
+  };
 
-  req.on('upgrade', (_res, _socket, _head) => {
-    callback();
-  });
+  const cleanup = (req, closeMethod) => {
+    if (!alreadyClosed) {
+      alreadyClosed = true;
+      setImmediate(() => { req[closeMethod](); });
+    }
+  };
+
+  const maybeRetry = (req, closeMethod, cond, ...args) => {
+    cleanup(req, closeMethod);
+    if (doCallback()) {
+      if (cond) {
+        retrying = true;
+        setImmediate(() => {
+          pingEndpoint(host, port, type, protocol, origin, callback, ++count);
+        });
+      } else {
```

Comment:
Could you remove this `if` with a `return` to limit the depth.

Reply:
I had a few ideas when I woke up this morning about how to improve control flow, will be making those now and will keep in mind opportunities for early `return`.

One thing, though, is that since `pingEndpoint` is doing asynchronous stuff and the return value is effectively ignored, there's no stack growth.

Reply:
A number of changes have been made. They've been squashed into the one commit for this PR, but if you want to see what changed specifically you can check https://github.com/embark-framework/embark/commit/8dfef2a66fba21b11e44a598cab9c809d13c217b — I was using that branch to test revised `pingEndpoint` and the proxy revisions together.

I tested on macOS, Windows, and desktop Linux, with websocket proxy and HTTP-only proxy. For all cases, I tried with `embark run` and `blockchain` in separate terminals and with single-terminal `embark run`, checking for blockchain process crashes when killing `run` and for hangs during startup. I tried jamming up the load on the CPU to see if I could get an indefinite stall... all very tedious, but I wanted to double-check everything so we can wrap this up with as much confidence as possible.

Everything worked correctly in all cases. Interestingly, on Linux I did *one* time see `Proxy: Network error 'read ECONNRESET'` reported in the output of `embark blockchain`. So I guess it's not just Windows on which that can happen.

I introduced an exponential increase in the timeout, with a max timeout of 3 seconds. As before, it will keep retrying for up to 10 minutes, after which the behavior becomes under defined as [described previously](https://github.com/embark-framework/embark/pull/1166#issuecomment-446413141).

I eliminated an edge case where a retry might be ongoing but another event might fire on a previous request. With my latest changes, if a request timeout has triggered a retry then additional events on the previous request will be ignored.

Some branching logic for handling error/success events has been combined into `maybeRetry` which is renamed `handleEvent`. An early `return` could be used [here](https://github.com/embark-framework/embark/blob/8dfef2a66fba21b11e44a598cab9c809d13c217b/src/lib/utils/utils.js#L132) and the `else if` branch would become an `if`; not sure if you have a strong preference on that change being made.

Finally, the changes re: the [HTTP timeout](https://github.com/embark-framework/embark/commit/8dfef2a66fba21b11e44a598cab9c809d13c217b#diff-1bdcf488d6e6476587e3166c3cf22177L171) were motivated by another look at the docs. The `request.setTimeout()` method doesn't take affect until after a socket connection has been made, which is the behavior we want so it's cleaner to use it than to setup a waterfall of event handlers.